File: frontend/src/types/agent.ts
// Agent runtime types, mirrored from backend SSE protocol.

export interface ToolCallState {
  callId: string
  toolName: string
  inputs: Record<string, unknown>
  output?: string
  isRunning: boolean
}

export interface SourceCitation {
  title: string
  chunk: string
  score: number
  documentId: string
}

export interface ChatUIMessage {
  id: string
  role: 'user' | 'assistant'
  content: string
  timestamp: number
  source?: 'entry' | 'agent' | 'system'
  thinking?: string
  toolCalls?: ToolCallState[]
  sources?: SourceCitation[]
  followUps?: string[]
  metadata?: Record<string, unknown> | null
  isStreaming?: boolean
}

export type SSEEventType =
  | 'stream_start'
  | 'agent_start'
  | 'message_delta'
  | 'thinking_delta'
  | 'tool_start'
  | 'tool_end'
  | 'follow_ups'
  | 'source_citations'
  | 'debug_timing'
  | 'agent_end'
  | 'stream_end'
  | 'error'

export interface AgentSession {
  id: string
  saas_agent_id: string
  user_id?: string | null
  title: string | null
  created_at: string
  updated_at: string
  message_count: number
}

export interface AgentMessageRow {
  id: string
  session_id: string
  role: string
  content: string
  tool_calls?: unknown
  thinking?: string | null
  sources?: Array<Record<string, unknown>> | null
  follow_ups?: string[] | null
  metadata?: Record<string, unknown> | null
  created_at: string
}

export interface AgentDocument {
  id: string
  saas_agent_id: string
  filename: string
  original_name: string
  content_type: string
  size_bytes: number
  chunk_count: number
  created_at: string
}

export interface AgentDocumentChunk {
  id: string
  document_id: string
  chunk_index: number
  content: string
  has_embedding: boolean
}

export interface AgentMemoryRow {
  id: string
  saas_agent_id: string
  session_id: string | null
  content: string
  category: string
  created_at: string
}

export interface AgentLearningCandidate {
  id: string
  saas_agent_id: string
  source_trace_id?: string | null
  trigger_type: string
  status: string
  title: string
  summary: string
  hint_text: string
  target_tool_name?: string | null
  target_action_path?: string | null
  target_risk_level?: string | null
  evidence?: Record<string, unknown> | null
  created_at: string
  reviewed_at?: string | null
}

export interface AgentHandoffContext {
  entry_session_id?: string | null
  saas_agent_id?: string | null
  saas_agent_name?: string | null
  entry_draft?: Record<string, unknown>
  connection_draft?: Record<string, unknown>
  active_connection_id?: string | null
  recent_entry_messages?: string[]
}

export interface AgentAdminStats {
  total_sessions: number
  total_messages: number
  total_documents: number
  total_memories: number
}

File: backend/core/schemas/agent.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class ChatRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=32_000)
    session_id: Optional[uuid.UUID] = None
    reasoning_mode: str = Field(default="balanced", pattern="^(fast|balanced|thorough)$")
    handoff_context: Optional[dict[str, Any]] = None


class AgentSessionRead(BaseModel):
    id: uuid.UUID
    saas_agent_id: uuid.UUID
    user_id: Optional[uuid.UUID] = None
    title: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    message_count: int = 0

    model_config = {"from_attributes": True}


class AgentSessionList(BaseModel):
    sessions: list[AgentSessionRead]
    total: int


class AgentMessageRead(BaseModel):
    id: uuid.UUID
    session_id: uuid.UUID
    role: str
    content: str
    tool_calls: Optional[list | dict] = None
    thinking: Optional[str] = None
    sources: Optional[list] = None
    follow_ups: Optional[list] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class AgentDocumentRead(BaseModel):
    id: uuid.UUID
    saas_agent_id: uuid.UUID
    filename: str
    original_name: str
    content_type: str
    size_bytes: int
    chunk_count: int
    created_at: datetime

    model_config = {"from_attributes": True}


class AgentDocumentChunkRead(BaseModel):
    id: uuid.UUID
    document_id: uuid.UUID
    chunk_index: int
    content: str
    has_embedding: bool = False

    model_config = {"from_attributes": True}


class AgentMemoryRead(BaseModel):
    id: uuid.UUID
    saas_agent_id: uuid.UUID
    session_id: Optional[uuid.UUID] = None
    content: str
    category: str
    created_at: datetime

    model_config = {"from_attributes": True}


class AgentMemoryCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=4000)
    category: str = Field(default="fact", pattern="^(fact|preference|instruction)$")


class AgentAdminStats(BaseModel):
    total_sessions: int
    total_messages: int
    total_documents: int
    total_memories: int


class AgentLearningCandidateRead(BaseModel):
    id: uuid.UUID
    saas_agent_id: uuid.UUID
    source_trace_id: Optional[uuid.UUID] = None
    trigger_type: str
    status: str
    title: str
    summary: str
    hint_text: str
    target_tool_name: Optional[str] = None
    target_action_path: Optional[str] = None
    target_risk_level: Optional[str] = None
    evidence: dict[str, Any] | None = None
    created_at: datetime
    reviewed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}
